### DISCUSSION TITLE
Limiting Ingest Variables (Bitrate, Resolution, Framerate etc)

### DISCUSSION THREAD

**Discussion by dbContext** [OvenMediaLabs/OvenMediaEngine#1090] [Ideas]
**Is your feature request related to a problem? Please describe.**

We utilise this in a production environment and for the majority of the time it's great, except when we have the occasional "cowboy" attempt to send us a very high bitrate/resolution because they can. Now on our busy days we're running 4/5+ concurrent streams so we limit only send out 640x360 @ 500Kbps as that is all that is needed, so all it's doing is wasting our transcode/ingest resources.

**Describe the solution you'd like**
Be able to limit the bitrate, resolution / framerate within the OME config, so let's say for example:

```
<Providers>
	<RTMP>
		<RestrictIngest>
			<Width>640</Width>
			<Height>360</Height>
			<Framerate>30.0</Framerate>
			<Bitrate>500000</Bitrate>
		</RestrictIngest>
	</RTMP>
</Providers>
```

If any of these variables are above the defined limits on initial RTMP ingest handshake, then the request is denied.

**Describe alternatives you've considered**

We already have a pre-generated OBS profile that creates the streaming defaults, to those stated above however again as the inherit problem suggests, it's due to overzealous IT guys trying to send us 1080p/2k/4k video because they can.

**Additional context**

Let me know on timeframe, providing it's something you guys would be interested in implementing - If the timeframe is long, I'll take a look at implementing it myself on the latest release and give you guys the chance to re-merge it.

**Comment by getroot** [comment]
This is a very good idea. And I think it's an essential feature in the real world.

This will be required for both MPEG-2 TS and RTSP as well as RTMP. This will also be required for future WebRTC providers. So I think this feature should be implemented in MediaRouter. Already MediaRouter is parsing Bitstream to extract the correct media profile. (RTMP's meta is inaccurate.)

Can you implement this related function in MediaRouter? We are now developing REST API, Recording, Thumbnail, RTMP Push, Signed URL. If you implement this feature we will be happy to merge it.

**Comment by dbContext** [comment]
@getroot Now 0.10.9 has been released, I'll start work on this.

**Comment by getroot** [comment]
SingedPolicy has been developed, but features to limit video input quality have not been developed yet. It will fit into the SingedPolicy specification.

**Comment by alexthedamager** [comment]
This is a very good idea, would likely be useful to many people 👍

[a comment by an automated account is not shown]

**Comment by Trungdo253** [comment]
Wow! It are exactly what i wanted. And can limiting ingest Codec? I'm waiting feauture in soon

**Comment by nadz-goldman** [comment]
One question: when ? =)

**Reply by bchah** [reply]
This can be achieved in a couple of ways now.

Firstly, check out the Alert module: https://docs.ovenmediaengine.com/alert

Or use AdmissionWebhooks in conjunction with the Stream Stats API to handle stream termination logic yourself: https://docs.ovenmediaengine.com/access-control/admission-webhooks
